// @gluestack/ui-next-adapter/dist/index.js
/**
 * @gluestack/ui-next-adapter
 * Next.js adapter for Gluestack UI
 */
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  withGluestackUI: () => withGluestackUI
});
module.exports = __toCommonJS(index_exports);

// src/utils.ts
var fs = __toESM(require("fs-extra"));
var path = __toESM(require("path"));
var isIncludedInDependency = (arr, inputString) => {
  return arr.includes(inputString);
};
function traverseFolder(dir, prefixes = [], deps = /* @__PURE__ */ new Set()) {
  const files = fs.readdirSync(dir);
  files.forEach((file) => {
    const filePath = path.join(dir, file);
    const stat = fs.statSync(filePath);
    if (stat.isDirectory()) {
      if (isIncludedInDependency(prefixes, file)) {
        deps.add(file);
      } else {
        traverseFolder(filePath, prefixes, deps);
      }
    }
  });
  return deps;
}
var getExactDependenciesFromNodeModules = (dir, prefixes = []) => {
  const nodeModulesDirectory = path.join(dir, "node_modules");
  const dependenciesSet = traverseFolder(nodeModulesDirectory, prefixes);
  const dependencyList = [];
  dependenciesSet.forEach((dependency) => {
    dependencyList.push(dependency);
  });
  return dependencyList;
};
function startsWithAny(string, array) {
  for (let i = 0; i < array.length; i++) {
    if (string.startsWith(array[i])) {
      return true;
    }
  }
  return false;
}
var getDependenciesFromNodeModules = (dir, nodeModulePackages = []) => {
  const myDependencies = /* @__PURE__ */ new Map();
  const traverse = (directory) => {
    const files = fs.readdirSync(directory);
    for (const file of files) {
      const filePath = path.join(directory, file);
      const stat = fs.statSync(filePath);
      if (stat.isDirectory()) {
        traverse(filePath);
      } else if (stat.isFile() && file === "package.json") {
        try {
          const packageJsonContent = fs.readFileSync(filePath, "utf8");
          const packageJson = JSON.parse(packageJsonContent);
          if (packageJson.name && startsWithAny(packageJson.name, nodeModulePackages)) {
            myDependencies.set(packageJson.name, {});
          }
        } catch (error) {
        }
      }
    }
  };
  nodeModulePackages.map((nodeModulePackage) => {
    const nodeModulesDirectory = path.join(
      dir,
      "node_modules",
      nodeModulePackage
    );
    if (fs.existsSync(nodeModulesDirectory)) {
      traverse(nodeModulesDirectory);
    }
  });
  const dependencyList = [];
  myDependencies.forEach((packageDependencies, packageName) => {
    dependencyList.push(packageName);
    Object.entries(packageDependencies).forEach(([dependencyName]) => {
      dependencyList.push(dependencyName);
    });
  });
  return dependencyList;
};
var checkIfWorkspace = (currDir) => {
  return checkIfWorkspaceRecursively(currDir);
};
var checkIfWorkspaceRecursively = (currDir) => {
  const parentFiles = fs.readdirSync(currDir);
  const metadata = {};
  if (parentFiles.includes("package.json")) {
    try {
      const packageJsonPath = path.resolve(currDir, "package.json");
      const packageJsonContent = fs.readFileSync(packageJsonPath, "utf8");
      const parentPackageJson = JSON.parse(packageJsonContent);
      const workspaces = parentPackageJson.workspaces;
      if (workspaces) {
        metadata["isWorkspace"] = true;
        metadata["workspaces"] = workspaces;
        metadata["workspacePath"] = currDir;
      }
    } catch (error) {
    }
  }
  const parentDir = path.resolve(currDir, "..");
  if (parentDir !== currDir) {
    const parentMetadata = checkIfWorkspaceRecursively(parentDir);
    if (parentMetadata.isWorkspace) {
      return parentMetadata;
    }
  }
  return metadata;
};

// src/withGluestack.ts
var import_find_yarn_workspace_root = __toESM(require("find-yarn-workspace-root"));
var gluestackDeps = [
  "@gluestack-ui",
  "@gluestack-ui",
  "@react-native-aria",
  "@gluestack",
  "@expo",
  "@legendapp",
  "expo-",
  "nativewind"
];
var reactNativeDeps = [
  "react-native",
  "react-native-web",
  "react-native-svg",
  "react-native-css-interop",
  "react-native-safe-area-context"
];
function withGluestackUI(nextConfig = {}) {
  var _a, _b, _c;
  const currDir = process.cwd();
  let rootDependencyList = [];
  let rootExactDependencyList = [];
  let parentDependencyList = [];
  let parentExactDependencyList = [];
  try {
    rootDependencyList = getDependenciesFromNodeModules(currDir, gluestackDeps);
  } catch (e) {
  }
  try {
    rootExactDependencyList = getExactDependenciesFromNodeModules(
      currDir,
      reactNativeDeps
    );
  } catch (e) {
  }
  const workspaceRoot = (0, import_find_yarn_workspace_root.default)(currDir);
  const metaWorkspace = checkIfWorkspace(currDir);
  if (metaWorkspace.isWorkspace) {
    try {
      parentDependencyList = getDependenciesFromNodeModules(
        metaWorkspace.workspacePath,
        gluestackDeps
      );
      parentExactDependencyList = getExactDependenciesFromNodeModules(
        metaWorkspace.workspacePath,
        reactNativeDeps
      );
    } catch (e) {
    }
  }
  if (workspaceRoot) {
    try {
      const workspaceRootDeps = getDependenciesFromNodeModules(
        workspaceRoot,
        gluestackDeps
      );
      const workspaceRootExactDeps = getExactDependenciesFromNodeModules(
        workspaceRoot,
        reactNativeDeps
      );
      parentDependencyList = [...parentDependencyList, ...workspaceRootDeps];
      parentExactDependencyList = [
        ...parentExactDependencyList,
        ...workspaceRootExactDeps
      ];
    } catch (e) {
    }
  }
  const allTranspileModules = Array.from(
    /* @__PURE__ */ new Set([
      ...rootDependencyList,
      ...parentDependencyList,
      ...rootExactDependencyList,
      ...parentExactDependencyList,
      ...nextConfig.transpilePackages || []
    ])
  );
  const baseConfig = {
    ...nextConfig,
    transpilePackages: allTranspileModules
  };
  const turboConfig = {
    resolveAlias: {
      ...((_a = nextConfig.turbopack) == null ? void 0 : _a.resolveAlias) || {},
      "react-native": "react-native-web"
    },
    resolveExtensions: [
      ".next15.js",
      ".next15.ts",
      ".next15.tsx",
      ".next15.jsx",
      ".web.js",
      ".web.jsx",
      ".web.ts",
      ".web.tsx",
      ".js",
      ".jsx",
      ".ts",
      ".tsx",
      ".json"
    ],
    rules: {
      ...((_b = nextConfig.turbopack) == null ? void 0 : _b.rules) || {},
      "*.{ttf,otf,woff,woff2}": {
        loaders: ["builtin:file-loader"],
        as: "*.js"
      }
    }
  };
  return {
    ...baseConfig,
    turbopack: {
      ...nextConfig.turbopack,
      ...turboConfig
    },
    // Keep experimental for any other experimental features
    experimental: {
      ...nextConfig.experimental
    },
    // Add env variables that work for both Webpack and Turbopack
    env: {
      ...nextConfig.env,
      DEV: process.env["NODE_ENV"] !== "production" ? "true" : "false",
      REACT_NATIVE_WEB: "true"
    },
    // Add compiler options for Turbopack (always include to ensure compatibility)
    compiler: {
      ...nextConfig.compiler,
      // Define global variables for Turbopack (similar to Webpack's DefinePlugin)
      define: {
        ...((_c = nextConfig.compiler) == null ? void 0 : _c.define) || {},
        "__DEV__": process.env["NODE_ENV"] !== "production" ? "true" : "false",
        "DEV": process.env["NODE_ENV"] !== "production" ? "true" : "false"
      }
    },
    webpack: (config, context) => {
      var _a2;
      if (nextConfig.webpack) {
        config = nextConfig.webpack(config, context);
      }
      config.resolve = config.resolve || {};
      config.resolve.alias = {
        ...config.resolve.alias || {},
        "react-native": "react-native-web"
      };
      config.resolve.extensions = [
        ".next15.js",
        ".next15.ts",
        ".next15.tsx",
        ".next15.jsx",
        ".web.js",
        ".web.jsx",
        ".web.ts",
        ".web.tsx",
        ...config.resolve.extensions
      ];
      config.module = config.module || {};
      config.module.rules = config.module.rules || [];
      config.module.rules.push({
        test: /\.(ttf|otf|woff|woff2)$/,
        use: {
          loader: "url-loader",
          options: {
            limit: 8192,
            fallback: "file-loader"
          }
        }
      });
      config.resolve.fallback = {
        ...config.resolve.fallback || {},
        "react-native-safe-area-context": false,
        "fs": false,
        "path": false
      };
      const { DefinePlugin } = require("webpack");
      const hasExistingDefines = (_a2 = config.plugins) == null ? void 0 : _a2.some(
        (plugin) => {
          var _a3, _b2;
          return plugin instanceof DefinePlugin && (((_a3 = plugin.definitions) == null ? void 0 : _a3.DEV) !== void 0 || ((_b2 = plugin.definitions) == null ? void 0 : _b2.__DEV__) !== void 0);
        }
      );
      if (!hasExistingDefines) {
        config.plugins = config.plugins || [];
        config.plugins.push(
          new DefinePlugin({
            "__DEV__": JSON.stringify(process.env["NODE_ENV"] !== "production"),
            "DEV": JSON.stringify(process.env["NODE_ENV"] !== "production"),
            "process.env.__DEV__": JSON.stringify(process.env["NODE_ENV"] !== "production"),
            "process.env.DEV": JSON.stringify(process.env["NODE_ENV"] !== "production"),
            "process.env.REACT_NATIVE_WEB": JSON.stringify("true")
          })
        );
      }
      return config;
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  withGluestackUI
});
